### PR TITLE
store: heap.Fix to update in O(log(n)), instead of 2 * O(log(n))

### PR DESCRIPTION
**Pull request by gyuho**
heap Fix, Push, Remove complexity are O(log(n)) where n = h.Len().
We should just call Fix once, O(log(n)) ,  instead of calling
Remove and Push which is 2 * O(log(n)). I also added tests.

### CODE REVIEW

**Comment by gyuho** [comment]
Just fixed govet error.

**Comment by gyuho** [comment]
Actually, there was `heap_test.go`. I will combine two.

**Comment by gyuho** [comment]
Changed the test file name to `ttl_key_heap_test.go`.

File: store/ttl_key_heap.go [hunk 1]
```diff
@@ -18,7 +18,8 @@ import (
 	"container/heap"
 )
 
-// An TTLKeyHeap is a min-heap of TTLKeys order by expiration time
+// An TTLKeyHeap is a min-heap of TTLKeys order by expiration time.
```
Comment (jonboulle) [inline store/ttl_key_heap.go]: An TTLKeyHeap --> ttlKeyHeap

Comment (gyuho) [inline store/ttl_key_heap.go]: Fixed. Thanks!

File: store/ttl_key_heap_test.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	th.update(nodesExpSoonestLast[4])
+
+	// update 7th element to third last
+	nodesExpSoonestLast[6].ExpireTime = time.Now().Add(time.Duration(heapN*8) * exp)
+	th.update(nodesExpSoonestLast[6])
+
+	// compare all the expiratonTime
+	i := 0
+	var expT time.Time
+	for th.Len() > 0 {
```
Comment (jonboulle) [inline store/ttl_key_heap_test.go]: `for i := 0; th.Len() > 0; i++ {`

File: store/ttl_key_heap_test.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+	// update 5th element to second last
+	nodesExpSoonestLast[4].ExpireTime = time.Now().Add(time.Duration(heapN*9) * exp)
+	th.update(nodesExpSoonestLast[4])
+
+	// update 7th element to third last
+	nodesExpSoonestLast[6].ExpireTime = time.Now().Add(time.Duration(heapN*8) * exp)
+	th.update(nodesExpSoonestLast[6])
+
+	// compare all the expiratonTime
```
Comment (jonboulle) [inline store/ttl_key_heap_test.go]: expireTime

**Comment by jonboulle** [comment]
Code change lgtm

File: store/ttl_key_heap_test.go (hunk 3 above)
Comment (gyuho) [inline store/ttl_key_heap_test.go]: Fixed. Thanks,

File: store/ttl_key_heap_test.go (hunk 2 above)
Comment (gyuho) [inline store/ttl_key_heap_test.go]: Fixed as well.

**Comment by xiang90** [comment]
I am not entirely sure if we still want to put more effort into store. It works well today and we do not really want it become scalable given the fact that we are focusing on v3 storage pkg.

**Comment by gyuho** [comment]
@xiang90 I understand. I was reading `store` code to understand more about `storage` package.

**Comment by xiang90** [comment]
@gyuho store and storage pkg are actually independent. Storage pkg is a self-contained pkg with very little external dependency.

**Comment by gyuho** [comment]
@xiang90 Makes sense. `store` pkg would still be used in v3? Or replaced by `storage`?

**Comment by xiang90** [comment]
@gyuho store pkg is only for v2. v3 will only use storage.

**Comment by gyuho** [comment]
@xiang90 Awesome! Thanks! I was working on other stuff in `store` package but I don't have to now. I will come back to this later, if we ever hit any performance hit. `watcherHub` uses `container/list`, and this can be replaced by slice or map if needed.

**Comment by xiang90** [comment]
The whole store thing is not scalable and it is designed for scalability. Other than just implementation issues, there are design issues too. That is why we have a completely new design for v3 in storage pkg.

**Comment by gyuho** [comment]
Got it. I will focus on `storage` then. Thanks,

**Pull request closed without merging**